### ISSUE TITLE
train_retriever_sup_unsup.json

### ISSUE THREAD

**Issue by chenzhongwu** [shuyanzhou/docprompting#19]
Could you share the code for generating the training data for simcse? What is the difference between text1 and text2 in train_retriever_sup_unsup.json?

**Comment by chenzhongwu** [comment]
I am very curious about how you generate the files in the provided dataset in https://drive.google.com/file/d/1CzNlo8-e4XqrgAME5zHEWEKIQMPga0xl/view?usp=sharing. What methods did you use to process them from what raw datasets?
Thanks!

**Comment by shuyanzhou** [comment]
> What is the difference between text1 and text2 in train_retriever_sup_unsup.json?

In the unsupervised setting, `text1` and `text2` are the same. In the supervised setting, `text2` is the natural language intent from `CoNaLa` and `text1` is the description of the function that fulfill the intent. 

> What methods did you use to process them from what raw datasets?

`CoNaLa` provides `NL-code` pairs. We use heuristics to extract the functions from the code and find their responding documents. Please see Appendix B of the [paper](https://arxiv.org/pdf/2207.05987) for more detailed descriptions. 

Let me know if you want to use similar pipelines to generate more `NL-doc-code` tuples. I can provide some more straightforward approach to generate the data.

**Comment by chenzhongwu** [comment]
Thanks! I have another question:  Could you explain more about Evaluation metrics: token-level F1 and character-level BLEU?

**Comment by chenzhongwu** [comment]
Thanks a lot! 
I have some other questions confusing me:  1. Could you explain more about Evaluation metrics: character-level BLEU? 2. What is the metric for Retrieval performance in Table 4? How to evaluate the retrieved docs are right? 3. Just curious: Why did you show the results under no-need-to-train BM25 in Table 1, instead of the other retrievers in Table 4?
Thanks again!

**Comment by chenzhongwu** [comment]
Thanks a lot! 
I have some other questions confusing me:  1. Could you explain more about Evaluation metrics: character-level BLEU? 2. What is the metric for Retrieval performance in Table 4? How to evaluate the retrieved docs are right?
Thanks again!

**Comment by shuyanzhou** [comment]
character-level BLEU is calculated in this way:
```python
from sacrebleu.metrics import BLEU
bleu = BLEU(tokenize='char')
bleu_score = bleu.corpus_score(pred_list, [src_list]).score
metric_list['bleu_char'] = bleu_score
```
where `pred_list` and `src_list` are `list[str]`

The code to calculate recall is [here](https://github.com/shuyanzhou/docprompting/blob/b7e0e626d47a0649090465bb15c91192fee8777f/retriever/eval.py#L110). Basically, we select `top-k` from the retriever and see if the ground-truth is inside the `top-k`.
